### PR TITLE
Fix ipv4_from_asc behavior on invalid Ipv4 addresses

### LINKED ISSUE #12649: a2i_IPADDRESS() considers "1.2.3.4.test.example" as valid IPv4 address
The function ``a2i_IPADDRESS()`` converts an ASCII IP address to an ``ASN1_OCTET_STRING`` and returns NULL when the input string is not a valid IPv4 or IPv6 address. In CPython I use the function to check if a user-supplied hostname is a DNS name or an IP address and then call either ``X509_VERIFY_PARAM_set1_host`` or ``X509_VERIFY_PARAM_set1_ip``. To my surprise the function also considers a hostname as valid IPv4 address if the input string starts with a valid IPv4 address. It ignores additional content. The ``sscanf()`` call in ``ipv4_from_asc`` does not check that the string is terminated immediately after the last digit.

https://github.com/openssl/openssl/blob/master/crypto/x509/v3_utl.c#L1088-L1093

Afftected Python ssl code:

https://github.com/python/cpython/blob/fa5d7251987c70a9c5d58b59a0b36ac9287eaafa/Modules/_ssl.c#L862-L925

Python bug report: https://bugs.python.org/issue41556

### PR DESCRIPTION
**Pull request by amiremohamadi**
Fixes #12649

### CODE REVIEW

**Comment by amiremohamadi** [comment]
Should I add some tests?

**Comment by beldmit** [comment]
Yes, it's worth doing. And also the tests failure seems relevant.

**Comment by amiremohamadi** [comment]
Thanks to @tiran, [here](https://github.com/openssl/openssl/pull/12651/files) is a draft test.
Can we use it?

**Comment by kaduk** [comment]
> Thanks to @tiran, [here](https://github.com/openssl/openssl/pull/12651/files) is a draft test.
> Can we use it?

I think it will be okay to use it.  Maybe someone will correct me later, but go ahead and incorporate it for now.

**Comment by t8m** [comment]
- This is definitely not a CLA: trivial change. Please remove the CLA: trivial from the commit message and submit a regular CLA - https://www.openssl.org/policies/cla.html
- Taking the test from another PR is IMO OK if that author has signed CLA already. However you need to keep the commit as is, i.e., keep the author and the commit message. You do that by fetching the tiran's branch and cherry-picking the commit into your branch.

**Comment by amiremohamadi** [comment]
Thanks @t8m. Some parts of the tests are different from what @tiran submitted. (Actually We expected different results on some cases). So should I cherry-pick the commit and apply my changes?

**Comment by t8m** [comment]
I'd cherry pick the commit and then apply your changes in a separate additional commit.

**Comment by richsalz** [comment]
> I'd cherry pick the commit and then apply your changes in a separate additional commit.

Why?  If the source PR is from someone who has a CLA, then anyone can build on top of their code for OpenSSL (or any other purpose).  See, for example, #15702 and #14464

**Comment by amiremohamadi** [comment]
> * This is definitely not a CLA: trivial change. Please remove the CLA: trivial from the commit message and submit a regular CLA - https://www.openssl.org/policies/cla.html
> * Taking the test from another PR is IMO OK if that author has signed CLA already. However you need to keep the commit as is, i.e., keep the author and the commit message. You do that by fetching the tiran's branch and cherry-picking the commit into your branch.

Done.
@t8m

**Review by t8m — APPROVED** [review]

File: crypto/x509/v3_utl.c [hunk 1]
```diff
@@ -1096,13 +1096,17 @@ int ossl_a2i_ipadd(unsigned char *ipout, const char *ipasc)
...
 static int ipv4_from_asc(unsigned char *v4, const char *in)
 {
-    int a0, a1, a2, a3;
+    const char *p;
+    int a0, a1, a2, a3, n;
 
-    if (sscanf(in, "%d.%d.%d.%d", &a0, &a1, &a2, &a3) != 4)
+    if (sscanf(in, "%d.%d.%d.%d%n", &a0, &a1, &a2, &a3, &n) != 4)
         return 0;
     if ((a0 < 0) || (a0 > 255) || (a1 < 0) || (a1 > 255)
```
Comment (paulidale) [inline crypto/x509/v3_utl.c:1104]: This condition is ugly....
On the other hand `(a0 | a1 | a2 | a3) & ~0377` is a thing of beauty.

No need to change.  Unless you agree :)

**Review by paulidale — APPROVED** [review]

File: crypto/x509/v3_utl.c (hunk 1 above)
Comment (amiremohamadi) [reply crypto/x509/v3_utl.c]: LGTM, changed.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: crypto/x509/v3_utl.c (hunk 1 above)
Comment (richsalz) [reply crypto/x509/v3_utl.c]: Oh dear no, that is a horrible ugly hack that nobody will understand the first time they look at the code.  As you had it before, it was clean and understandable.  Please don't.

Comment (t8m) [reply crypto/x509/v3_utl.c]: Maybe adding a comment `/* all numbers must be within 0 and 255 */` would make it OK even for you @richsalz?

Comment (t8m) [reply crypto/x509/v3_utl.c]: @amiremohamadi could you please add that comment?

Comment (richsalz) [reply crypto/x509/v3_utl.c]: I'm not on the project. I don't get why you're trying to save a couple of clock cycles.  Shrug.

Comment (davidben) [reply crypto/x509/v3_utl.c]: +1 to @richsalz's comment. Write the obvious thing. It won't matter and, besides, the compiler can figure this out for you:
https://godbolt.org/z/r4n1rr6Tf

Comment (tiran) [reply crypto/x509/v3_utl.c]: Would are more restrictive format like ``%hhu`` work or is it not portable / undefined for leading minus sign?

```
    unsigned char a0, a1, a2, a3;
    int n;

    if (sscanf(in, "%hhu.%hhu.%hhu.%hhu%n", &a0, &a1, &a2, &a3, &n) != 4)
```

**Comment by t8m** [comment]
In the end I've decided to merge only the first three commits as the condition change is controversial. Thank you for the contribution.

**Comment by kaduk** [comment]
> In the end I've decided to merge only the first three commits as the condition change is controversial. Thank you for the contribution.

My reading was that the spelling of the condition change was controversial, but the behavior change itself was agreed upon as fine.

**Comment by t8m** [comment]
> > In the end I've decided to merge only the first three commits as the condition change is controversial. Thank you for the contribution.
> 
> My reading was that the spelling of the condition change was controversial, but the behavior change itself was agreed upon as fine.

I am talking here about 8a2b872. This commit does not introduce any behavior change. It only optimizes and obfuscates the existing condition which was disputed as something undesirable.

**Pull request closed without merging**